src/main/java/com/cloudbees/groovy/cps/impl/CpsClosure.java

```diff
@@ -36,5 +43,15 @@ public Object call(Object arguments) {
         throw new CpsCallableInvocation(def,this,arguments);
     }
 
+    /**
+     * {@link ClosureWriter} generates this function with actual argument types.
+     * Here we approximate by using varargs.
+     * <p>
+     * {@link CurriedClosure} invokes this method directly (via {@link MetaClassImpl#invokeMethod(Class, Object, String, Object[], boolean, boolean)}
+     */
+    public Object doCall(Object... args) {
```

Comment:
The next step up from here is to have `CpsTransformer` produce AST that defines a new `CpsClosure` subtypes that have the correct type signature in `doCall` method. Not something I want to do unless the need is proven.